pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -501,12 +501,90 @@ func (e *DockercfgController) createDockerPullSecret(serviceAccount *v1.ServiceA
 		return nil, false, err
 	}
 	dockercfgSecret.Data[v1.DockerConfigKey] = dockercfgContent
+	dockercfgSecret.SetOwnerReferences([]metav1.OwnerReference{
+		{
+			Kind:       "Secret",
+			APIVersion: "v1",
+			Name:       tokenSecret.Name,
+			UID:        tokenSecret.UID,
+		},
+	})
 
 	// Save the secret
 	createdSecret, err := e.client.CoreV1().Secrets(tokenSecret.Namespace).Create(dockercfgSecret)
 	return createdSecret, err == nil, err
 }
 
+func (e *DockercfgController) syncDockercfgOwnerRefs(serviceAccount *v1.ServiceAccount) error {
+	for _, secretRef := range serviceAccount.Secrets {
+		secret, exists, err := e.secretCache.GetByKey(fmt.Sprintf("%s/%s", secretRef.Namespace, secretRef.Name))
+		if err != nil {
+			return err
+		}
+		if !exists {
+			continue
+		}
+		err = e.syncDockercfgOwner(secret.(*v1.Secret))
+		if err != nil {
+			return err
+		}
+	}
+	for _, secretRef := range serviceAccount.ImagePullSecrets {
+		secret, exists, err := e.secretCache.GetByKey(fmt.Sprintf("%s/%s", serviceAccount.Namespace, secretRef.Name))
+		if err != nil {
+			return err
+		}
+		if !exists {
+			continue
+		}
+		err = e.syncDockercfgOwner(secret.(*v1.Secret))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *DockercfgController) syncDockercfgOwner(pullSecret *v1.Secret) error {
+	if pullSecret.Type != v1.SecretTypeDockercfg {
+		return nil
+	}
+	tokenName := pullSecret.Annotations[ServiceAccountTokenSecretNameKey]
+	// If there is no token name, this pull secret was likely linked by a user.
+	// No further work is needed.
+	if len(tokenName) == 0 {
+		return nil
+	}
+	// Make sure the token exists
+	tokenSecret, exists, err := e.secretCache.GetByKey(fmt.Sprintf("%s/%s", pullSecret.Namespace, tokenName))
+	if err != nil {
+		return err
+	}
+	if !exists {
+		// If the pull secret exists, and the token does not, delete the pull secret.
+		klog.V(4).Infof("Deleting pull secret %s/%s because its associated token %s/%s is missing", pullSecret.Namespace, pullSecret.Name, pullSecret.Namespace, tokenName)
+		return e.client.CoreV1().Secrets(pullSecret.Namespace).Delete(pullSecret.Name, &metav1.DeleteOptions{})
+	}
+	tokenSecretObj := tokenSecret.(*v1.Secret)
+	pullSecret = pullSecret.DeepCopy()
```

Comment:
minor optimization, it seems safe to make this copy after you check if the ownerref already exists, so you can skip doing the copy most of the time (since normally the ref will exist and you're done)